### PR TITLE
Add RedisResult.AsBoolSlice() #499

### PR DESCRIPTION
**Pull request by Ali-Assar**
Hi. I added .AsBoolSlice() as requested

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -743,6 +753,30 @@ func (m *RedisMessage) AsFloatSlice() ([]float64, error) {
 	return s, nil
 }
 
+// AsBoolSlice checks if the message is a Redis array/set response, and converts it to []bool.
+// Redis nil elements and other non-boolean elements will be represented as false.
+func (m *RedisMessage) AsBoolSlice() ([]bool, error) {
+	if m.typ != '*' {
+		panic(fmt.Sprintf("redis message type %c is not an array", m.typ))
+	}
```
Comment (rueian) [inline message.go:761]:
```suggestion
```

This check is already done by `m.ToArray()`

File: message.go [hunk 2]
```diff
@@ -743,6 +753,30 @@ func (m *RedisMessage) AsFloatSlice() ([]float64, error) {
...
+	s := make([]bool, len(values))
+	for i, v := range values {
+		if len(v.string) != 0 {
+			s[i], err = strconv.ParseBool(v.string)
+			if err != nil {
+				s[i] = false
+			}
+		} else {
+			s[i] = v.integer != 0
+		}
```
Comment (rueian) [inline message.go:775]:
```suggestion
		s[i], _ = v.AsBool()
```
We can simply re-use `.AsBool()` here and ignore its error.

File: message_test.go [hunk 3]
```diff
@@ -292,6 +292,27 @@ func TestRedisResult(t *testing.T) {
...
+			t.Fatal("AsBoolSlice not failed as expected")
+		}
+		values := []RedisMessage{{string: "true", typ: '+'}}
+		if ret, _ := (RedisResult{val: RedisMessage{typ: '*', values: values}}).AsBoolSlice(); !reflect.DeepEqual(ret, []bool{true}) {
+			t.Fatal("AsBoolSlice not get value as expected")
+		}
+		values = []RedisMessage{{string: "false", typ: '+'}}
+		if ret, _ := (RedisResult{val: RedisMessage{typ: '*', values: values}}).AsBoolSlice(); !reflect.DeepEqual(ret, []bool{false}) {
+			t.Fatal("AsBoolSlice not get value as expected")
+		}
```
Comment (rueian) [inline message_test.go:309]: true/false strings are good use cases. Could you make `AsBool()` support them as well?

File: message_test.go [hunk 4]
```diff
@@ -1384,6 +1405,26 @@ func TestRedisMessage(t *testing.T) {
...
+	})
+
+	defer func() {
+		if r := recover(); r == nil {
+			t.Fatal("AsBoolSlice did not panic as expected")
+		} else if !strings.Contains(r.(string), "redis message type t is not an array") {
+			t.Fatal("AsBoolSlice did not panic with the expected message")
+		}
+	}()
+	(&RedisMessage{typ: 't'}).AsBoolSlice()
```
Comment (rueian) [inline message_test.go:1426]:
```suggestion
```
duplicated.

File: message.go [hunk 5]
```diff
@@ -743,6 +753,20 @@ func (m *RedisMessage) AsFloatSlice() ([]float64, error) {
...
+// AsBoolSlice checks if the message is a redis array/set response, and converts it to []bool.
+// Redis nil elements and other non-boolean elements will be present as false.
+func (m *RedisMessage) AsBoolSlice() ([]bool, error) {
+	values, err := m.ToArray()
+	if err != nil {
+		return nil, err
+	}
+	s := make([]bool, len(values))
+	for i, v := range values {
+		s[i] = v.string == "1"
```
Comment (rueian) [inline message.go:765]: I think your original use of `strconv.ParseBool` was good. Why did you remove it completely? I believe we better add it to AsBool() and re-use AsBool() here.

**Comment by Ali-Assar** [comment]
This time I used AsBool function within AsBoolSlice and I think this approach is fine and better than strconv.ParseBool as the logic for parsing a boolean value is centralized and consistent across the codebase.

Please let me now if there is any other feedback

**Review by rueian — APPROVED** [review]
LGTM
